### PR TITLE
make C++ types hashable

### LINKED ISSUE #87: C++ types should be hashable
Currently, there is no implementation of `std::hash<>` for `renf_elem_class`. There probably should be one so `renf_elem_class` can be used in `std::unordered_set/map`.

### PR DESCRIPTION
**Pull request by saraedum**
fixes #87

### CODE REVIEW

**Comment by saraedum** [comment]
@videlec what do you think about this naive hash function? Is double conversion guaranteed to produce the correct double (with whatever rounding) or do I need to to refine first to make this reliable?

**Comment by videlec** [comment]
The `get_d` function is not guaranteed to give you the closest floating point. The problem is with balls like `0.999999999 ± 0.000000001` which are not taken care of carefully. Why do you want to use the floating point value for hashing? Why not the `fmpq` coefficients?

**Comment by saraedum** [comment]
Are the coefficients compatible with `==`? Currently we do not support comparing elements in different number fields but it would not work then anymore. Already, rationals would need special treatment so that they hash is compatible with the `==` operator. Finally, it's a bit more complicated to combine coefficients into a hash value.

Maybe it's easier to fix `get_d`?

**Comment by videlec** [comment]
Computationally, `get_d` is likely to be more expansive. Fixing `get_d` would also be a good thing. At C level, we could provide two hashes (e.g. `renf_elem_hash_coeffs` and `renf_elem_hash_d`). That might help choosing one for the default C++ one.

**Comment by saraedum** [comment]
`get_d` is expensive. But maybe there's something that can be done about it so the call after the first is not expensive?

I don't mind implementing `renf_elem_hash_coeffs` and `renf_elem_hash_d` but if the former is not compatible with `operator==` then I don't see why we would want to use it in C++.

**Comment by saraedum** [comment]
@videlec: I added some tests for get_d. Debugging what's going wrong, I think I found a problem in arb. I'll create an issue there.

What do you think about the `fail.h` I added here? I think writing tests is currently quite cumbersome. Would you mind if we started using some library for this that makes writing tests a bit easier?

**Comment by saraedum** [comment]
@videlec this is ready I think, assuming that the CI is happy now.

I included the changes to the testing framework indicated in #116. It is a bit weird to test C code from C++ but I found it really tedious not to use a testing framework. We could of course use a C framework but the we would have two different testing frameworks here (and another two more for benchmarks?)

All the generator code is boilerplate that only has to be written once™ so we might not want to worry about it too much.

File: renf_elem/test/t-get_d.cpp [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    arf_clear(a);
+}
+
+TEST_CASE("Convert renf_elem to double", "[renf_elem][get_d]") {
+    flint_rand_t& state = GENERATE(rands());
+    renf_t& nf = GENERATE_REF(renfs(state));
+    renf_elem_t& a = GENERATE_REF(renf_elems(state, nf));
+
+    CAPTURE(nf);
+    CAPTURE(a);
```
Comment (saraedum) [inline renf_elem/test/t-get_d.cpp:55]: `CAPTURE` is a macro. `nf` and `a` will be printed when a test fails:
```
../e-antic/renf_elem/test/t-get_d.cpp:52: FAILED:
due to a fatal error condition:
  nf := renf_t(emb=[-1.
  8678980663715787745372895523767659458063827519418234750099407562112074968152-
  1028251688189134472090784125535016161591108453600257131449109834971723823471-
  3558517486795191772222970157636285937468473899051911457798116405515449012948-
  2184880266849315660741726641050277671291024383371868545103992846169525202625-
  9091999719390053027148006374567593251590308345142506859456022204431364608871-
  2002020491768596316073184281277208422377354767087096577102943378324273789408
  +/- 8.30e-457], pol=1135594640*x^20 + 3195475668*x^19 + 11677172927*x^18 +
  9934829224*x^17 - 16316074178*x^16 + 1518116458*x^15 + 12140149242*x^14 +
  1678026534*x^13 - 13559455876*x^12 + 11057028813*x^11 + 16422010505*x^10 +
  6410129766*x^9 + 6301606743*x^8 + 8699492605*x^7 + 6989347111*x^6 +
  13347687583*x^5 + 17894147445*x^4 + 14082144089*x^3 + 13459114366*x^2 +
  17556251915*x + 8368987156, prec=1253)
  a := -8825*x^19 - 210180*x^18 - 4096*x^17 + 4*x^16 + 44906*x^15 - 493100*x^13
  - 49*x^12 + 16227*x^11 - 89*x^10 - 7092*x^9 + 147*x^7 - 126*x^5 + 391*x^4 -
  3*x^3 - 162518*x ~ [-1.355328010e+10 +/- 2.71]
  SIGINT - Terminal interrupt signal
```

File: renf_elem/test/t-get_d.cpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    fmpz_clear(man);
+    fmpz_clear(exp);
+    fmpz_clear(quo);
+    arf_clear(a);
+}
+
+TEST_CASE("Convert renf_elem to double", "[renf_elem][get_d]") {
+    flint_rand_t& state = GENERATE(rands());
+    renf_t& nf = GENERATE_REF(renfs(state));
+    renf_elem_t& a = GENERATE_REF(renf_elems(state, nf));
```
Comment (saraedum) [inline renf_elem/test/t-get_d.cpp:52]: Unfortunately it's a bit complicated to bring these pure C objects into the lifetime assumptions made by the catch framework. So we need these `&` to make things works which are, I think, the only thing here that does not actually look like C++ code.

File: renfxx/test/main.cpp [hunk 3]
```diff
@@ -0,0 +1,2 @@
+#define CATCH_CONFIG_MAIN
+#include "external/catch2/single_include/catch2/catch.hpp"
```
Comment (saraedum) [inline renfxx/test/main.cpp:2]: This gets linked with every test that uses catch and provides the main for it, i.e., the command line interface.

File: renf_elem/test/t-get_d.cpp (hunk 2 above)
Comment (saraedum) [reply renf_elem/test/t-get_d.cpp:52]: The code here iterates through 128 `renf_t`s and for each one through 128 `renf_elem_t`s. These calls can be further adjusted with optional parameters, e.g., the first optional parameter is the number of iterations.

File: renf_elem/test/t-get_d.cpp (hunk 1 above)
Comment (saraedum) [reply renf_elem/test/t-get_d.cpp:55]: (here I pressed Ctrl-C.)

File: renf_elem/test/t-get_d.cpp (hunk 2 above)
Comment (videlec) [reply renf_elem/test/t-get_d.cpp:52]: How do you know it is 128 of them?

Comment (saraedum) [reply renf_elem/test/t-get_d.cpp:52]: It's the default in that generator. If you search the code for 128 you should find two hits.

**Comment by videlec** [comment]
clang is very picky
```
renf_elem/get_d.c:136:40: error: comparing floating point with == or != is unsafe [-Werror,-Wfloat-equal]
            if (arf_get_d(ubound, rnd) == d)
                ~~~~~~~~~~~~~~~~~~~~~~ ^  ~
1 error generated.
```

**Review by videlec — APPROVED** [review]

**Comment by videlec** [comment]
thanks

**Pull request merged**